### PR TITLE
3DCursor: Adding gearwear input device

### PR DESCRIPTION
**Pull request by parthmehta209**
The gearwear directory contains necessary projects
to enable the use of a Samsung GearS2/Gear2 as an IoDevice in the
3DCursorLibrary

GearInputConsumer:
    Has the tizen project for the app on the watch side.

PhoneSide/gearinputprovider:
    Implements the provider side for the GearInputConsumer app on the watch.

PhoneSide/gearwearlibrary:
    Library that any app can use to get events from the watch.

GearWearIoDevice:
    Implements an IoDevice as defined by the 3DCursorLibrary using the
    gearwearlibrary

### CODE REVIEW

**Comment by rahul27** [comment]
@parthmehta209 is it possible to replace some of the unmodified source files into a pre-compiled library? would make the code more manageable and readable.

You could provide to link to the source and instructions to build in case someone wishes to modify it.

**Comment by parthmehta209** [comment]
@rahul27 I removed the themes from tau library used in GearInputConsumer. That brings down the total files to 68 for the entire implementation.

File: GVRf/Extensions/3DCursor/IODevices/gearwear/GearInputConsumer/js/main.js [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+ */
+// TODO: Extract constants, figure out why formatting is breaking up 'const varName' into 2 lines
+// TODO: Exit application on page hide (home button press) since back press is overriden
+// TODO: Proper lifecycle of network (on show/hide)
+// TODO: See how connect/disconnect lifecycle should be. Is there a way to automatically connect/disconnect (when watch connects/disconnects from phone)
+// TODO: See if we can use WIFI as well
+// TODO: Better structure for sending json
+// TODO: If network not connected, don't try sending events
+// TODO: Don't show too many popups if network not connected
+var MAIN = (function main() {
```
Comment (rahul27) [inline GVRf/Extensions/3DCursor/IODevices/gearwear/GearInputConsumer/js/main.js]: Better to have logs turned on using a debug flag.

File: GVRf/Extensions/3DCursor/IODevices/gearwear/GearWearIoDevice/src/main/java/com/gearvrf/io/gearwear/GearWearableDevice.java [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+import org.gearvrf.io.cursor3d.CustomKeyEvent;
+import org.gearvrf.io.cursor3d.IoDevice;
+import org.gearvrf.utility.Log;
+
+import java.util.concurrent.TimeUnit;
+
+// TODO: Consumer needs to send event when application exits (unless TouchEnd will occur)
+// TODO: Need to detect connection issues, and stop moving if issue occurs
+// TODO: See if new algorithm works well on Gear 2, since it has a 320x320 square display
+public class GearWearableDevice extends IoDevice {
```
Comment (rahul27) [inline GVRf/Extensions/3DCursor/IODevices/gearwear/GearWearIoDevice/src/main/java/com/gearvrf/io/gearwear/GearWearableDevice.java]: Have the logs turned on using a debug flag.

File: GVRf/Extensions/3DCursor/IODevices/gearwear/PhoneSide/gearinputprovider/src/main/java/com/samsung/mpl/gearinputprovider/backend/InputProviderService.java [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.lang.ref.WeakReference;
+import java.util.HashMap;
+
+public class InputProviderService extends SAAgent {
+    private static final String TAG = InputProviderService.class.getSimpleName();
+    private static final String HELLO_WORLD_LINKED_PACKAGE_NAME = "com.samsung.smcl" +
```
Comment (rahul27) [inline GVRf/Extensions/3DCursor/IODevices/gearwear/PhoneSide/gearinputprovider/src/main/java/com/samsung/mpl/gearinputprovider/backend/InputProviderService.java]: rename to input provider service.

File: GVRf/Extensions/3DCursor/IODevices/gearwear/PhoneSide/gearinputprovider/src/main/java/com/samsung/mpl/gearinputprovider/utils/NotificationUtility.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+import android.graphics.Bitmap.CompressFormat;
+import android.graphics.BitmapFactory;
+
+import com.samsung.mpl.gearinputprovider.R;
+import com.samsung.mpl.gearinputprovider.models.NotificationModel;
+
+import java.io.ByteArrayOutputStream;
+
+public class NotificationUtility {
+    private static final String GEAR_WIDGET_ID = "VbsgE6bms3.HelloWorldConsumer";
```
Comment (rahul27) [inline GVRf/Extensions/3DCursor/IODevices/gearwear/PhoneSide/gearinputprovider/src/main/java/com/samsung/mpl/gearinputprovider/utils/NotificationUtility.java]: Could we rename the watch app?

**Comment by rahul27** [comment]
You can also remove GVRf/Extensions/3DCursor/IODevices/gearwear/GearInputConsumer/img/gear_sample.jpg.

File: GVRf/Extensions/3DCursor/IODevices/gearwear/GearWearIoDevice/src/main/java/com/gearvrf/io/gearwear/GearWearableDevice.java (hunk 2 above)
Comment (parthmehta209) [inline GVRf/Extensions/3DCursor/IODevices/gearwear/GearWearIoDevice/src/main/java/com/gearvrf/io/gearwear/GearWearableDevice.java]: done

File: GVRf/Extensions/3DCursor/IODevices/gearwear/GearInputConsumer/js/main.js (hunk 1 above)
Comment (parthmehta209) [inline GVRf/Extensions/3DCursor/IODevices/gearwear/GearInputConsumer/js/main.js]: done

File: GVRf/Extensions/3DCursor/IODevices/gearwear/PhoneSide/gearinputprovider/src/main/java/com/samsung/mpl/gearinputprovider/backend/InputProviderService.java (hunk 3 above)
Comment (parthmehta209) [inline GVRf/Extensions/3DCursor/IODevices/gearwear/PhoneSide/gearinputprovider/src/main/java/com/samsung/mpl/gearinputprovider/backend/InputProviderService.java]: removed that part of the code all together. It was was two way communication, and currently only one way communication (from watch to phone) is used.

File: GVRf/Extensions/3DCursor/IODevices/gearwear/PhoneSide/gearinputprovider/src/main/java/com/samsung/mpl/gearinputprovider/utils/NotificationUtility.java (hunk 4 above)
Comment (parthmehta209) [inline GVRf/Extensions/3DCursor/IODevices/gearwear/PhoneSide/gearinputprovider/src/main/java/com/samsung/mpl/gearinputprovider/utils/NotificationUtility.java]: removed the file.

**Comment by rahul27** [comment]
Just setting up my Gear S2 to test this patch. Should be good to go once I verify the sample app.

**Comment by parthmehta209** [comment]
@rahul27 Added a Readme.txt file with instructions to build and run the watch app with the phone app.

**Comment by liaxim** [comment]
@rahul27 Can this be merged?

**Comment by rahul27** [comment]
I still need to test it, Couldn't get my S2 up and running. Will try again today.

**Comment by rahul27** [comment]
Verified this and it works.

**Pull request merged**